tsdb/shard.go

```diff
@@ -1736,73 +1732,127 @@ func (fs *MeasurementFieldSet) IsEmpty() bool {
+// and start the background write process
+func (fs *MeasurementFieldSet) SetMeasurementFieldSetWriter(queueLength int) {
+	fs.writer = &MeasurementFieldSetWriter{writeErrChannels: make(chan errorChannel, queueLength)}
+	go fs.saveWriter()
+}
 
-	// Write the new index to a temp file and rename when it's sync'd
-	// if it is still the most recent memoryVersion of the MeasurementFields
-	path := fs.path + "." + strconv.FormatUint(v, 10) + ".tmp"
+func (fs *MeasurementFieldSet) Save() error {
+	return fs.writer.RequestSave()
+}
 
-	fd, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR|os.O_EXCL|os.O_SYNC, 0666)
-	if err != nil {
-		return err
-	}
-	defer os.RemoveAll(path)
+func (w *MeasurementFieldSetWriter) RequestSave() error {
+	done := make(chan error)
+	w.writeErrChannels <- done
+	return <-done
+}
 
-	if _, err := fd.Write(fieldsIndexMagicNumber); err != nil {
-		return err
+func (fs *MeasurementFieldSet) saveWriter() {
+	for {
+		// Block until someone modifies the MeasurementFieldSet and
+		// it needs to be written to disk.
+		errorChannels, err := fs.writeToFile(<-fs.writer.writeErrChannels)
+		for _, c := range errorChannels {
+			c <- err
+			close(c)
+		}
 	}
+}
 
-	if _, err := fd.Write(b); err != nil {
-		return err
-	}
+// writeToFile: Write the new index to a temp file and rename when it's sync'd
+func (fs *MeasurementFieldSet) writeToFile(first errorChannel) (errorChannels []errorChannel, err error) {
+	// Put the errorChannel on which we blocked into a slice to allow more invocations
+	// to share the return code from the file write
+	errorChannels = append(errorChannels, first)
 
-	if err = fd.Sync(); err != nil {
-		return err
+	// Do some blocking IO operations before marshalling the in-memory index
+	// to allow other changes to it to be queued up and be captured in one
+	// write operation, in case we are under heavy field creation load
+	path := fs.path + ".tmp"
+
+	// Open the temp file
+	fd, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR|os.O_EXCL|os.O_SYNC, 0666)
+	if err != nil {
+		return errorChannels, err
 	}
+	// Ensure temp file is cleaned up
+	defer func() {
+		if e := os.RemoveAll(path); err == nil {
+			err = e
+		}
+	}()
+	isEmpty, err := func() (isEmpty bool, err error) {
+		// ensure temp file closed before rename (for Windows)
+		defer func() {
+			if e := fd.Close(); err == nil {
+				err = e
+			}
+		}()
+		if _, err = fd.Write(fieldsIndexMagicNumber); err != nil {
+			return true, err
+		}
 
-	//close file handle before renaming to support Windows
-	if err = fd.Close(); err != nil {
-		return err
+		// Read all the pending new field and measurement write requests
+		// that will be captured in the marshaling of the in-memory copy
+		for {
+			select {
+			case ec := <-fs.writer.writeErrChannels:
+				errorChannels = append(errorChannels, ec)
+				continue
+			default:
+			}
+			break
+		}
+		// Lock, copy, and marshal the in-memory index
+		b, err := fs.marshalMeasurementFieldSet()
```

Comment:
Why not re-check for pending writes here? If there's a new write, we could throw away the marshalled data and re-marshal.

Do we have any perf test setup that we could check which way would be better?

Reply:
I didn't recheck because the copying and marshaling locks the structure, preventing new changes to it, so the window for new updates to capture is small, at the cost of significant code complexity. My assumption was that the file open and initial write was a nice blocking IO operation that would let other threads run and possibly modify the MeasurementFieldSet for subsequent capture by the marshalMeasurementFieldSet